### PR TITLE
Global dump

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | yes |
| Applies to | all |
| Fixed tickets | - |

### CODE REVIEW

File: components/var_dumper/introduction.rst [hunk 1]
```diff
@@ -62,6 +62,15 @@ current PHP SAPI:
...
     wherever you want.
 
+.. tip::
+
+    In order to have the ``dump()`` function always available, you can install
+    it globally on your dev environment. Just follow these steps: run
+    ``composer global require symfony/var-dumper`` then add
+    ``auto_prepend_file = ${HOME}/.composer/vendor/autoload.php``
+    to your ``php.ini`` file. Don't forget to run ``composer global update``
+    from time to time to keep up with the latest fixes.
```
Comment (WouterJ) [inline components/var_dumper/introduction.rst]: I think it's more clear to put this in a step list:

````rst
.. tip::

    In order to have to ``dump()`` function always available, you can install
    it globally in your dev environment:

    #. Run ``composer global require symfony/var-dumper``;
    #. Add ``auto_prepend_file = ${HOME}/.composer/vendor/autoload.php``
       to your ``php.ini`` file;
    #. (Don't forget to run ``composer global update`` from time to time to always
       have the latest bug fixes.)
````

Comment (nicolas-grekas) [inline components/var_dumper/introduction.rst]: Fixed, thanks

File: components/var_dumper/introduction.rst [hunk 2]
```diff
@@ -62,6 +62,17 @@ current PHP SAPI:
...
     wherever you want.
 
+.. tip::
+
+    In order to have the ``dump()`` function always available, you can install
+    it globally in your dev environment:
+
+    #. Run ``composer global require symfony/var-dumper``;
+    #. Add ``auto_prepend_file = ${HOME}/.composer/vendor/autoload.php``
+       to your ``php.ini`` file;
```
Comment (xabbuh) [inline components/var_dumper/introduction.rst]: Does PHP interpret environment variables in the `php.ini` file? What happens if the file doesn't exist?

**Comment by WouterJ** [comment]
Looks like a very nice tip to add to the documentation! :)

File: components/var_dumper/introduction.rst (hunk 2 above)
Comment (nicolas-grekas) [inline components/var_dumper/introduction.rst]: 1. Yes it does: http://php.net/manual/en/configuration.file.php
2. A fatal error

Comment (xabbuh) [inline components/var_dumper/introduction.rst]: Should we add a warning then? Otherwise, people could get these errors when they execute PHP scripts with different users.

Comment (nicolas-grekas) [inline components/var_dumper/introduction.rst]: I think the error message is good enough to not clutter the doc, what do you think?

**Comment by lyrixx** [comment]
:+1:

File: components/var_dumper/introduction.rst [hunk 3]
```diff
@@ -62,6 +62,17 @@ current PHP SAPI:
     You'll also learn how to change the format or redirect the output to
     wherever you want.
 
+.. tip::
+
+    In order to have the ``dump()`` function always available, you can install
+    it globally in your dev environment:
```
Comment (javiereguiluz) [inline components/var_dumper/introduction.rst]: The phrase `... in your dev environment`  may sound confusing. This has nothing to do with the `dev` environment of Symfony applications. I would replace it by:

```
In order to have the ``dump()`` function available for any PHP script or application and not only for your Symfony applications, you can install it globally in your machine:
```

Comment (WouterJ) [inline components/var_dumper/introduction.rst]: I do agree with you, but I also don't like the verbose thing you proposed. The docs are always a fight between making something easy to understand and keeping something as short as possible (especially when it's about simple tips like this).

What do you think about replacing "environment" with "machine" or "box"?

File: components/var_dumper/introduction.rst [hunk 4]
```diff
@@ -62,6 +62,17 @@ current PHP SAPI:
...
+
+    #. Run ``composer global require symfony/var-dumper``;
+    #. Add ``auto_prepend_file = ${HOME}/.composer/vendor/autoload.php``
+       to your ``php.ini`` file;
+    #. (Don't forget to run ``composer global update`` from time to time to always
```
Comment (javiereguiluz) [inline components/var_dumper/introduction.rst]: With the new list layout, I would remove the parenthesis of this item. And I would reword it as follows:

```
Execute the ``composer global update`` command from time to time to
update this component and always have the latest bug fixes.
```

Comment (WouterJ) [inline components/var_dumper/introduction.rst]: I don't think "update this component" adds much extra value.

I've added the parenthesis because it's not really a next step in the installation process, it's just something you have to do in the time after it. Another option is to add it as a normal paragraph after the list

File: components/var_dumper/introduction.rst (hunk 3 above)
Comment (javiereguiluz) [inline components/var_dumper/introduction.rst]: It's true that my proposal is more verbose, but I think it explains the purpose of the tip better.

In the original text, what does *"to have the function always available"*? That's why I prefer to say explicitly that this is referring to using it outside Symfony applications.

And the same for `dev environment` or `dev box`. Prefer to say `your machine`, `your computer`, `your local machine`, etc.

File: components/var_dumper/introduction.rst (hunk 4 above)
Comment (javiereguiluz) [inline components/var_dumper/introduction.rst]: I agree with the last part. As you said, this is not a point of the list ... so maybe it shouldn't be on the list after all.

**Comment by nicolas-grekas** [comment]
Thanks for the comments guys, PR updated. Ok for you?

**Comment by stof** [comment]
I see an issue with this proposal: the composer global project can contain lots of other stuff as well, which might create issues in your project. To do it in a safe way, you should do a composer installation where you only have the var-dumper component

File: components/var_dumper/introduction.rst [hunk 5]
```diff
@@ -62,6 +62,17 @@ current PHP SAPI:
...
 
+.. tip::
+
+    In order to have the ``dump()`` function always available when running
+    any PHP code, you can install it globally on your computer:
+
+    #. Run ``composer global require symfony/var-dumper``;
+    #. Add ``auto_prepend_file = ${HOME}/.composer/vendor/autoload.php``
+       to your ``php.ini`` file;
+    #. From time to time, run ``composer global update`` to have the latest
```
Comment (WouterJ) [inline components/var_dumper/introduction.rst]: @javiereguiluz and I concluded to put this after the list (not as a list item). If you agree, could you please do that? (I hope it's the last fix, as you've had to push way to many times for a simple tip like this)

**Comment by WouterJ** [comment]
@stof that's exactly what I was thinking. Or maybe we have to recommend to add `$HOME/.composer/vendor/symfony/var-dumper/Resources/functions/dump.php` as auto prepend file.

(or create a custom auto prepend file in the PHP dir which requires this one)

**Comment by nicolas-grekas** [comment]
@stof lets document an easy way for new comers. Advanced users that know what they do with composer will find their way. New comers won't have many "composer global" things (it's not a very common thing, or you're an advanced user).
This just works (c)

File: components/var_dumper/introduction.rst (hunk 5 above)
Comment (nicolas-grekas) [inline components/var_dumper/introduction.rst]: I prefer it with a 3d item (I talked last with @javiereguiluz offline, so I win ;-) )

**Comment by nicolas-grekas** [comment]
For the reader wondering about autoloading precedence, global classes are always looked up *after* project class. So, the project always takes precedence.

File: components/var_dumper/introduction.rst (hunk 5 above)
Comment (WouterJ) [inline components/var_dumper/introduction.rst]: Okay, you one. :+1: then :)

**Comment by Pierstoval** [comment]
Absolutely awesome, I've just installed it and it works like a charm, this is REALLY great ! :)

File: components/var_dumper/introduction.rst (hunk 2 above)
Comment (xabbuh) [inline components/var_dumper/introduction.rst]: I'm not sure. Will each and every beginner really understand that a global Composer installation is only global in regard of their user account? We should keep in mind that the web server might not be the same account as the one the developer is using.

**Comment by weaverryan** [comment]
@nicolas-grekas how did you know that I was wondering about autoloading precedence? :) Because of that, I like this.

Thanks!
